src/beliefs/discrete.jl

```diff
@@ -1,133 +1,107 @@
+        for j = 1:length(b)
+            s += alphas[j,i]*b[j]
+        end
+        util[i] = s
     end
-    b
+    return util
 end
 
-function Base.setindex!(b::DiscreteBelief, x::Float64, i::Int64) 
-    b.b[i] = x
-    b
-end
 
-Base.getindex(b::DiscreteBelief, i::Int64) = b.b[i]
 
-function Base.copy!(b1::DiscreteBelief, b2::DiscreteBelief)
-    copy!(b1.b, b2.b)
-    return b1
+mutable struct DiscreteUpdater{P<:POMDP} <: Updater
+    pomdp::P
 end
 
-Base.sum(b::DiscreteBelief) = sum(b.b)
-
-create_belief(bu::DiscreteUpdater) = DiscreteBelief(n_states(bu.pomdp))
+# Will this cause excessive calls to DiscreteBelief ?
+create_belief(bu::DiscreteUpdater) = DiscreteBelief(bu.pomdp)
 
-function initialize_belief(bu::DiscreteUpdater, dist::Any, belief::DiscreteBelief = create_belief(bu))
+function initialize_belief(bu::DiscreteUpdater, dist::Any, belief::DiscreteBelief=create_belief(bu))
     belief = fill!(belief, 0.0)
     for s in iterator(dist)
-        sidx = state_index(bu.pomdp, s) 
-        belief[sidx] = pdf(dist, s)  
+        sidx = state_index(bu.pomdp, s)
+        belief.b[sidx] = pdf(dist, s)
     end
     return belief
 end
 
+function update(bu::DiscreteUpdater, b::DiscreteBelief, a, o)
+    pomdp = b.pomdp
+    state_space = b.state_list
+    bp = zeros(length(state_space))
+
+    bp_sum = 0.0   # to normalize the distribution
+
+    for (spi, sp) in enumerate(state_space)
 
-# Updates the belief given the current action and observation
-function update{A,O}(bu::DiscreteUpdater, bold::DiscreteBelief, a::A, o::O)
-    bnew = create_belief(bu)
-    pomdp = bu.pomdp
-    # initialize spaces
-    pomdp_states = ordered_states(pomdp)
-    # ensure belief state sizes match 
-    @assert length(bold) == length(bnew)
-    # initialize belief 
-    fill!(bnew, 0.0)
-    # iterate through each state in belief vector
-    for (i, sp) in enumerate(pomdp_states)
-        # get the distributions
+        # po = O(a, sp, o)
         od = observation(pomdp, a, sp)
-        # get prob of observation o from current distribution
-        probo = pdf(od, o)
-        # if observation prob is 0.0, then skip rest of update b/c bnew[i] is zero
-        if probo == 0.0
+        po = pdf(od, o)
+
+        if po == 0.0
             continue
         end
-        b_sum = 0.0 # belief for state sp
-        for (j, s) in enumerate(pomdp_states)
+
+        b_sum = 0.0
+        for (si, s) in enumerate(state_space)
             td = transition(pomdp, s, a)
             pp = pdf(td, sp)
-            b_sum += pp * bold[j]
+            b_sum += pp * b.b[si]
         end
-        bnew[i] = probo * b_sum
+
+        bp[spi] = po * b_sum
+        bp_sum += bp[spi]
     end
-    norm = sum(bnew)
-    # if norm is zero, the update was invalid - reset to uniform
-    if norm == 0.0
-        println("Invalid update for: ", bold, " ", a, " ", o)
-        u = 1.0/length(bnew)
-        fill!(bnew, u)
+
+    if bp_sum == 0.0
+        error("INVALID BELIEF UPDATE")
```

Comment:
Can we make this error more informative. Even something like
```julia
error("""
Failed discrete belief update: new probabilities sum to zero.

b = $b
a = $a
o = $o
""")
```
is much more helpful - the user will then know that the observation function is outputting a distribution with zeros where they shouldn't be.

Reply:
Definitely! I should point out that the previous version did not throw an error for belief crash; instead it reinitialized the belief (uniformly, I think?). However, I feel strongly that if you experience belief crash something terrible has gone wrong and you probably have issues with your observation function that need to be fixed.

Reply:
Although printing out the whole belief is a bit iffy... in many cases it will be a huge list and the user might not even see the text, "Failed discrete belief update..."